### PR TITLE
name async generators something more human friendly in type error diagnostic

### LINKED ISSUE #81457: Diagnostic involving the type of an async block refers to `generator`'s
version 1.49:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6cf5c01427a274fed3395916ea86ab87

`FuturesUnordered` requires futures with the same type:
```
use futures::stream::FuturesUnordered;
fn main() {

    let u = FuturesUnordered::new();
    
    u.push(async {});
    u.push(async {});
    
}
```

results in:
```
error[E0308]: mismatched types
  --> src/main.rs:12:18
   |
11 |     u.push(async {});
   |                  -- the expected generator
12 |     u.push(async {});
   |                  ^^ expected generator, found a different generator
   |
   = note: expected generator `[static generator@src/main.rs:11:18: 11:20]`
              found generator `[static generator@src/main.rs:12:18: 12:20]`
```

As far as I know, generators have 2 types in their internal representation, so their display impl (its not actually display as far as I remember, but whatever creates the string to describe a type), should me more accurate about it being a future

also, these spans seem off?

@rustbot label +A-async-await +A-diagnostics +D-papercut

### PR DESCRIPTION
**Pull request by guswynn**
fixes #81457

Some details:

1. I opted to load the generator kind from the hir in TyCategory. I also use 1 impl in the hir for the descr
2. I named both the source of the future, in addition to the general type (`future`), not sure what is preferred
3. I am not sure what is required to make sure "generator" is not referred to anywhere. A brief `rg "\"generator\"" showed me that most diagnostics correctly distinguish from generators and async generator, but the `descr` of `DefKind` is pretty general (not sure how thats used) 
4. should the descr impl of AsyncGeneratorKind use its display impl instead of copying the string?

### CODE REVIEW

File: src/test/ui/issues/issue-81457.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// edition:2018
+use std::future::Future;
+
+fn fun<F: Future<Output = ()>>(f1: F, f2: F) {}
+fn main() {
+    fun(async {}, async {});
```
Comment (oli-obk) [inline src/test/ui/issues/issue-81457.rs:6]: Please also test the other kinds of async blocks

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -1319,6 +1328,16 @@ impl fmt::Display for AsyncGeneratorKind {
     }
 }
 
+impl AsyncGeneratorKind {
+    pub fn descr(&self) -> &'static str {
+        match self {
+            AsyncGeneratorKind::Block => "future (`async` block)",
+            AsyncGeneratorKind::Closure => "future (`async` closure body)",
+            AsyncGeneratorKind::Fn => "future (`async fn` body)",
```
Comment (oli-obk) [inline compiler/rustc_hir/src/hir.rs:1336]: looking at the actual error message I think we should leave out the `future` and parentheses and just mention the `async` block and similar part. The fact that it's a future is technically not relevant for the type of the thing

**Review by oli-obk — CHANGES_REQUESTED** [review]
What do you think about the change in diagnostics (leaving out the `future` part?)

File: compiler/rustc_hir/src/hir.rs (hunk 2 above)
Comment (guswynn) [reply compiler/rustc_hir/src/hir.rs:1336]: In my experience, people think of all these things as "futures", as that's the bound on every api that takes multiple futures of the same type (FuturesUnordered is how I hit this). Though I need to add a test where I try to use an async block and an async fn in the same api and see what it looks like

Comment (oli-obk) [reply compiler/rustc_hir/src/hir.rs:1336]: hmm... once you have an `impl Future`, you will get errors about opaque types and not the ones you created. This is similar to how closures have a "closure type", but once you have them in an `impl Fn` you only get an error about opaque types.

As a more concrete example, look at the error message in this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1363fce8006b6776f59c0192ace4ddc

Talking about a "future" here is similar to talking about a "callable" in the case of closures. The actual type of the closure is an anonymous closure type.

Comment (guswynn) [reply compiler/rustc_hir/src/hir.rs:1336]: Hmm this is a good point. when I find time this weekend I'll add more tests. (Also the notes in those errors messages are nice, I should add something like that, and perhaps a note about these async generators being futures)

**Comment by JohnCSimon** [comment]
Ping from triage - @guswynn 
can you please resolve the build failure?

@Rustbot label: -S-waiting-on-review +S-waiting-on-author

**Comment by guswynn** [comment]
Yah, I will have time for this tomorrow!

**Comment by guswynn** [comment]
Fixed the build issue (apparently ui/issues is too big these days?)

also, @oli-obk It appears that async fn's trigger a different path with arguably better (tho somewhat confusing) wording, and async closures also have better wording

I think the async fn one could be improved, but as a separate pr. I have left it in the test because we should have a test for the 3 main async generator types regardless

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/82262#issuecomment-781440859)

@bors r-

**Comment by guswynn** [comment]
@Dylan-DPC looks like I got unlucky with a merge that made hir.rs larger...I added a commit with the suppression the error suggested

**Pull request merged**